README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# swiping_card_deck

[![Flutter Tests](https://img.shields.io/github/actions/workflow/status/jushutch/swiping_card_deck/badges.yml?branch=main&label=Tests&style=flat-square)](https://github.com/jushutch/swiping_card_deck/tree/main/test)
[![Test Coverage](https://img.shields.io/codecov/c/github/jushutch/swiping_card_deck/main?label=Coverage&style=flat-square)](https://app.codecov.io/gh/jushutch/swiping_card_deck)
[![Bugs](https://img.shields.io/github/issues/jushutch/swiping_card_deck/bug?label=Bug%20Issues&style=flat-square)](https://github.com/jushutch/swiping_card_deck/labels/bug)

A widget for swiping through a deck of cards with gestures or buttons.

This package was inspired when I was trying to develop a Tinder-like app
that involved swiping options to the left or right. I assumed that a package
existed for such a popular functionality, but I searched the internet and could
only find bits and pieces of different implementations that were confusing,
complicated, and down right bad. This package aims to be an easy-to-use and
customizable way to implement that functionality, and will hopefully save you
a great amount of time!

![SwipingCardDeck Demonstration](https://github.com/jushutch/swiping_card_deck/raw/main/gif/swiping_card_deck.gif)

## Features

The SwipingCardDeck widget offers a variety of unique features:

- Swipe through cards by dragging or using buttons to hook into the public
  swipeLeft and swipeRight functions.
- Provide custom callback functions for when the deck is empty and for each
  swiping direction, which also pass in the original Card object.
- Use custom Cards of any shape, size, or content.
- Optimized performance by rendering top two cards at a time, allowing for large
  decks with no decrease in user experience.
- A variety of exposed properties allow for a swiping experience that can be
  designed for any need.
- Great for any Tinder-like decision making application.

To create a deck of any widget, use the generic SwipingDeck class with templating!

What this package doesn't support:

- Vertical swiping

## Usage

Constructing a SwipingCardDeck with two cards and two buttons that can be used
for swiping. Callback functions print out debugging information. The parameters
minimumVelocity, rotationFactor, swipeThreshold, and swipeAnimationDuration are all optional, but are set
to the default values which work well for most use cases. The disableDragging parameter is set to false by
default, but can be set to true in order to prevent cards from being swiped using the dragging gesture.

```dart
SwipingCardDeck(
    cardDeck: <Card>[
        Card(
            color: Colors.red,
            child: const SizedBox(height: 300, width: 200,)
        ),
        Card(
            color: Colors.green,
            child: const SizedBox(height: 300, width: 200,)
        ),
    ],
    onDeckEmpty: () => debugPrint("Card deck empty"),
    onLeftSwipe: (Card card) => debugPrint("Swiped left!"),
    onRightSwipe: (Card card) => debugPrint("Swiped right!"),
    swipeThreshold: MediaQuery.of(context).size.width / 4,
    minimumVelocity: 1000,
    cardWidth: 200,
    rotationFactor: 0.8 / 3.14,
    swipeAnimationDuration: const Duration(milliseconds: 500),
    disableDragging: false,
);
```

## Additional information

### Issues and suggestions

This package is being actively maintained and developed. To submit ideas,
issues, or suggestions, create an issue in the
[GitHub repository](https://github.com/jushutch/swiping_card_deck).

### Contributing

Thank you for your interest in contributing! To get started, read through
the documentation in [CONTRIBUTING](https://github.com/jushutch/swiping_card_deck/blob/main/CONTRIBUTING.md).
There are always new issues coming so be sure to check back often, and if there's
something you want to work on but there's no issue, just open one yourself!


CHANGELOG.md
## 1.3.0

- Add option to disable swiping by dragging

## 1.2.0

- Create generic SwipingDeck for swiping through other widgets

## 1.1.0

- Expose the duration of the swiping animation

## 1.0.0

- Initial release! :tada:
- **BREAKING**: Remove screenSize parameter from swipe hooks
- Add unit tests
- Fix bug with swipeController late initialization
- Add badges to README

## 0.0.3

- Fix GIF issue in README

## 0.0.2

- Update README
- Add unit tests
- Move SwipingGestureDetector into src
- Fix issue with animation disposal

## 0.0.1

- Initial package release.


test/swiping_card_deck_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:swiping_card_deck/swiping_card_deck.dart';
import 'dart:math' as math;

void main() {
  const int numCards = 5;
  List<Card> getMockCards() {
    List<Card> cardDeck = [];
    for (int i = 0; i < numCards; ++i) {
      cardDeck.add(
        Card(
            color: Color((math.Random().nextDouble() * 0xFFFFFF).toInt())
                .withOpacity(1.0),
            child: const SizedBox(height: 300, width: 200)),
      );
    }
    return cardDeck;
  }

  final cardDeck = getMockCards();

  Future<void> _mountWidget(WidgetTester tester) async {
    SwipingCardDeck mockDeck = SwipingCardDeck(
      cardDeck: cardDeck,
      onDeckEmpty: () => debugPrint("Card deck empty"),
      onLeftSwipe: (dynamic card) => debugPrint("Swiped left!"),
      onRightSwipe: (dynamic card) => debugPrint("Swiped right!"),
      cardWidth: 200,
    );
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: mockDeck,
      ),
    ));
  }

  testWidgets("SwipingCardDeck widget is created", (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder deckFinder = find.byType(SwipingCardDeck);
    expect(deckFinder, findsOneWidget);
  });

  testWidgets("Only two cards are rendered", (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder cardFinder = find.byType(Card);
    expect(cardFinder, findsNWidgets(2));
  });

  testWidgets("swipeLeft removes top card and runs callback",
      (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder cardFinder = find.byType(Card);
    expect(cardFinder, findsNWidgets(2));
    expect(tester.widget(cardFinder.first) as Card, cardDeck[1]);
    expect(tester.widget(cardFinder.last) as Card, cardDeck[0]);

    Finder deckFinder = find.byType(SwipingCardDeck);
    expect(deckFinder, findsOneWidget);

    SwipingCardDeck deck = tester.widget(deckFinder);
    expectLater(() => deck.swipeLeft(), prints("Swiped left!\n"));
    await tester.pumpAndSettle();
    expect(cardFinder, findsNWidgets(2));
    expect(tester.widget(cardFinder.first) as Card, cardDeck[2]);
    expect(tester.widget(cardFinder.last) as Card, cardDeck[1]);
  });

  testWidgets("swipeRight removes top card and runs callback",
      (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder deckFinder = find.byType(SwipingCardDeck);
    expect(deckFinder, findsOneWidget);
    SwipingCardDeck deck = tester.widget(deckFinder);
    Finder cardFinder = find.byType(Card);
    expect(cardFinder, findsNWidgets(2));
    expect(tester.widget(cardFinder.first) as Card, cardDeck[1]);
    expect(tester.widget(cardFinder.last) as Card, cardDeck[0]);

    expectLater(() => deck.swipeRight(), prints("Swiped right!\n"));
    await tester.pumpAndSettle();
    expect(cardFinder, findsNWidgets(2));
    expect(tester.widget(cardFinder.first) as Card, cardDeck[2]);
    expect(tester.widget(cardFinder.last) as Card, cardDeck[1]);
  });

  testWidgets("Callback function is ran when deck is empty",
      (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder deckFinder = find.byType(SwipingCardDeck);
    expect(deckFinder, findsOneWidget);
    SwipingCardDeck deck = tester.widget(deckFinder);

    for (int i = 0; i < numCards - 1; ++i) {
      deck.swipeLeft();
      await tester.pumpAndSettle();
    }

    Finder cardFinder = find.byType(Card);
    expect(cardFinder, findsOneWidget);
    expect(tester.widget(cardFinder.first) as Card, cardDeck[numCards - 1]);

    expectLater(
        () => deck.swipeLeft(), prints("Swiped left!\nCard deck empty\n"));
    await tester.pumpAndSettle();

    expect(cardFinder, findsNothing);
  });
}


test/swiping_gesture_detector_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:swiping_card_deck/src/swiping_gesture_detector.dart';
import 'dart:math' as math;

void main() {
  const int numCards = 5;
  List<Card> getMockCards() {
    List<Card> cardDeck = [];
    for (int i = 0; i < numCards; ++i) {
      cardDeck.add(
        Card(
          color: Color((math.Random().nextDouble() * 0xFFFFFF).toInt())
              .withOpacity(1.0),
          child: const SizedBox(height: 300, width: 200),
        ),
      );
    }
    return cardDeck;
  }

  Future<void> _mountWidget(WidgetTester tester) async {
    final List<Card> cardDeck = getMockCards();
    SwipingGestureDetector mockDetector = SwipingGestureDetector(
      cardDeck: cardDeck,
      swipeLeft: () {
        debugPrint("Swiped left!");
        cardDeck.removeLast();
      },
      swipeRight: () {
        debugPrint("Swiped right!");
        cardDeck.removeLast();
      },
      cardWidth: 200,
      swipeThreshold: 200,
      minimumVelocity: 1000,
      rotationFactor: 0,
    );
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: mockDetector,
      ),
    ));
  }

  Future<void> _mountDisabledWidget(WidgetTester tester) async {
    final List<Card> cardDeck = getMockCards();
    SwipingGestureDetector mockDetector = SwipingGestureDetector(
      cardDeck: cardDeck,
      swipeLeft: () {
        debugPrint("Swiped left!");
        cardDeck.removeLast();
      },
      swipeRight: () {
        debugPrint("Swiped right!");
        cardDeck.removeLast();
      },
      cardWidth: 200,
      swipeThreshold: 200,
      minimumVelocity: 1000,
      rotationFactor: 0,
      disableDragging: true,
    );
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: mockDetector,
      ),
    ));
  }

  testWidgets("SwipingGestureDetector widget is created",
      (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder gestureFinder = find.byType(SwipingGestureDetector);
    expect(gestureFinder, findsOneWidget);
  });

  testWidgets("Card is swiped at correct thresholds",
      (WidgetTester tester) async {
    await _mountWidget(tester);
    Finder gestureFinder = find.byType(SwipingGestureDetector);
    expect(gestureFinder, findsOneWidget);
    SwipingGestureDetector detector = tester.widget(gestureFinder);

    // Test the right swipe threshold
    await expectLater(
        () async => await tester.drag(gestureFinder, const Offset(200, 0)),
        prints("Swiped right!\n"));
    expect(detector.cardDeck.length, numCards - 1);
    await expectLater(
        () async => await tester.drag(gestureFinder, const Offset(199, 0)),
        prints(""));
    expect(detector.cardDeck.length, numCards - 1);

    // Test the left swipe threshold
    await expectLater(
        () async => await tester.drag(gestureFinder, const Offset(-200, 0)),
        prints("Swiped left!\n"));
    expect(detector.cardDeck.length, numCards - 2);
    await expectLater(
        () async => await tester.drag(gestureFinder, const Offset(-199, 0)),
        prints(""));
    expect(detector.cardDeck.length, numCards - 2);
  });

  testWidgets("Card springs back to center", (WidgetTester tester) async {
    // NOTE: The Offset values for the drag gesture are rounded because
    // the dragAlignment is converted from pixels to a percentage which
    // is imprecise.
    await _mountWidget(tester);
    Finder gestureFinder = find.byType(SwipingGestureDetector);
    expect(gestureFinder, findsOneWidget);
    SwipingGestureDetector detector = tester.widget(gestureFinder);
    Finder topCardFinder = find.byWidget(detector.cardDeck.last);
    expect(topCardFinder, findsOneWidget);
    Offset initial = tester.getCenter(topCardFinder);

    // Run drag gesture that moves to the right 30, holds, and then releases
    TestGesture gesture = await tester.startGesture(initial);
    await gesture.moveBy(const Offset(30, 0));
    await tester.pumpAndSettle();

    Offset newCenter = tester.getCenter(topCardFinder);
    newCenter = Offset(
        newCenter.dx.round().toDouble(), newCenter.dy.round().toDouble());
    expect(newCenter, initial + const Offset(30, 0));

    // Assert that the top card is animated after the drag is released
    await gesture.up();
    await tester.pump(const Duration(milliseconds: 2));
    expect(tester.getCenter(topCardFinder), isNot(newCenter));
    expect(tester.getCenter(topCardFinder), isNot(initial));

    // Assert that the top card returns to the initial position
    await tester.pumpAndSettle();
    newCenter = tester.getCenter(topCardFinder);
    newCenter = Offset(
        newCenter.dx.round().toDouble(), newCenter.dy.round().toDouble());
    expect(newCenter, initial);
  });

  testWidgets("Card is swiped at minimum velocity",
      (WidgetTester tester) async {
    // NOTE: The fling velocity values are slightly generous due to the
    // imprecision of the drag details velocity.
    await _mountWidget(tester);
    Finder gestureFinder = find.byType(SwipingGestureDetector);
    expect(gestureFinder, findsOneWidget);
    SwipingGestureDetector detector = tester.widget(gestureFinder);

    // Test the right swipe velocity
    await expectLater(
        () async =>
            await tester.fling(gestureFinder, const Offset(199, 0), 1001),
        prints("Swiped right!\n"));
    expect(detector.cardDeck.length, numCards - 1);
    await expectLater(
        () async =>
            await tester.fling(gestureFinder, const Offset(199, 0), 999),
        prints(""));
    expect(detector.cardDeck.length, numCards - 1);
    await tester.pumpAndSettle();

    // Test the left swipe velocity
    await expectLater(
        () async =>
            await tester.fling(gestureFinder, const Offset(-199, 0), 1001),
        prints("Swiped left!\n"));
    expect(detector.cardDeck.length, numCards - 2);
    await expectLater(
        () async =>
            await tester.fling(gestureFinder, const Offset(-199, 0), 999),
        prints(""));
    expect(detector.cardDeck.length, numCards - 2);
  });

  testWidgets("Card cannot be swiped when dragging is disabled",
      (WidgetTester tester) async {
    // NOTE: The fling velocity values are slightly generous due to the
    // imprecision of the drag details velocity.
    await _mountDisabledWidget(tester);
    Finder gestureFinder = find.byType(SwipingGestureDetector);
    expect(gestureFinder, findsOneWidget);
    SwipingGestureDetector detector = tester.widget(gestureFinder);

    // Test flinging to the right
    await expectLater(
        () async =>
            await tester.fling(gestureFinder, const Offset(199, 0), 1001),
        prints(""));
    expect(detector.cardDeck.length, numCards);
    await tester.pumpAndSettle();

    // Test flinging to the left
    await expectLater(
        () async =>
            await tester.fling(gestureFinder, const Offset(-199, 0), 1001),
        prints(""));
    expect(detector.cardDeck.length, numCards);

    // Test the right swipe threshold
    await expectLater(
        () async => await tester.drag(gestureFinder, const Offset(200, 0)),
        prints(""));
    expect(detector.cardDeck.length, numCards);

    // Test the left swipe threshold
    await expectLater(
        () async => await tester.drag(gestureFinder, const Offset(-200, 0)),
        prints(""));
    expect(detector.cardDeck.length, numCards);
  });
}


lib/swiping_card_deck.dart
/// The swiping_card_deck library contains the [SwipingCardDeck] widget
/// which allows a user to swipe through a deck of [Card] widgets. To
/// swipe through any [Widget], use the generic [SwipingDeck] class.
library swiping_card_deck;

import 'package:flutter/material.dart';
import './src/swiping_gesture_detector.dart';

/// A [SwipingDeck] of [Card] widgets
typedef SwipingCardDeck = SwipingDeck<Card>;

/// A deck of [Widget] objects that can be swiped to the left or right
/// using a gesture or a button.
// ignore: must_be_immutable
class SwipingDeck<T extends Widget> extends StatelessWidget {
  SwipingDeck(
      {Key? key,
      required this.cardDeck,
      required this.onLeftSwipe,
      required this.onRightSwipe,
      required this.onDeckEmpty,
      required this.cardWidth,
      this.minimumVelocity = 1000,
      this.rotationFactor = .8 / 3.14,
      this.swipeThreshold,
      this.swipeAnimationDuration = const Duration(milliseconds: 500),
      this.disableDragging = false})
      : super(key: key) {
    cardDeck = cardDeck.reversed.toList();
  }

  /// The list of [Widget] objects to be swiped.
  List<T> cardDeck;

  /// Callback function ran when a [Widget] is swiped left.
  final Function(T) onLeftSwipe;

  /// Callback function ran when a [Widget] is swiped right.
  final Function(T) onRightSwipe;

  /// Callback function when the last [Widget] in the [cardDeck] is swiped.
  final Function() onDeckEmpty;

  /// The minimum horizontal velocity required to trigger a swipe.
  final double minimumVelocity;

  /// The amount each [Widget] rotates as it is swiped.
  final double rotationFactor;

  /// The width of all [Widget] objects in the [cardDeck].
  final double cardWidth;

  /// The [Duration] of the swiping [AnimationController].
  final Duration swipeAnimationDuration;

  /// Disable swiping cards using the dragging gesture.
  final bool disableDragging;

  /// The distance in pixels that a [Widget] must be dragged before it is swiped.
  late final double? swipeThreshold;

  /// The [SwipingGestureDetector] used to control swipe animations.
  late final SwipingGestureDetector swipeDetector;

  /// The [Size] of the screen.
  late final Size screenSize;

  bool animationActive = false;
  static const String left = "left";
  static const String right = "right";

  @override
  Widget build(BuildContext context) {
    screenSize = MediaQuery.of(context).size;
    swipeDetector = SwipingGestureDetector(
      cardDeck: cardDeck,
      swipeLeft: swipeLeft,
      swipeRight: swipeRight,
      swipeThreshold: swipeThreshold ?? screenSize.width / 4,
      cardWidth: cardWidth,
      rotationFactor: rotationFactor,
      minimumVelocity: minimumVelocity,
      swipeAnimationDuration: swipeAnimationDuration,
      disableDragging: disableDragging,
    );
    return SizedBox(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          SizedBox(
            width: MediaQuery.of(context).size.width,
            child: swipeDetector,
          ),
        ],
      ),
    );
  }

  /// Swipe the top [Widget] to the left.
  ///
  /// If there is no animation already in progress, trigger the animation
  /// to swipe the top [Widget] to the left, call the function [onLeftSwipe],
  /// and remove the [Widget] from the deck. If the deck is empty, call the
  /// function [onDeckEmpty].
  Future<void> swipeLeft() async {
    if (animationActive || cardDeck.isEmpty) return;
    await _swipeCard(left, screenSize);
    onLeftSwipe(cardDeck.last);
    cardDeck.removeLast();
    if (cardDeck.isEmpty) onDeckEmpty();
  }

  /// Swipe the top [Widget] to the right.
  ///
  /// If there is no animation already in progress, trigger the animation
  /// to swipe the top [Widget] to the right, call the function [onRightSwipe],
  /// and remove the [Widget] from the deck. If the deck is empty, call the
  /// function [onDeckEmpty].
  Future<void> swipeRight() async {
    if (animationActive || cardDeck.isEmpty) return;
    await _swipeCard(right, screenSize);
    onRightSwipe(cardDeck.last);
    cardDeck.removeLast();
    if (cardDeck.isEmpty) onDeckEmpty();
  }

  Future<void> _swipeCard(String direction, Size screenSize) async {
    double swipeEnd = screenSize.width / 2 + cardWidth;
    swipeDetector.swipe = AlignmentTween(
      begin: swipeDetector.dragAlignment,
      end: Alignment(direction == left ? -swipeEnd : swipeEnd,
          swipeDetector.dragAlignment.x),
    ).animate(
      CurvedAnimation(
        parent: swipeDetector.swipeController,
        curve: Curves.ease,
      ),
    );
    animationActive = true;
    await swipeDetector.swipeController.forward();
    swipeDetector.swipeController.reset();
    animationActive = false;
  }
}


lib/src/swiping_gesture_detector.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:flutter/physics.dart';

//ignore: must_be_immutable
class SwipingGestureDetector<T> extends StatefulWidget {
  SwipingGestureDetector({
    Key? key,
    required this.cardDeck,
    required this.swipeLeft,
    required this.swipeRight,
    required this.cardWidth,
    this.minimumVelocity = 1000,
    this.rotationFactor = .8 / 3.14,
    this.swipeAnimationDuration = const Duration(milliseconds: 500),
    required this.swipeThreshold,
    this.disableDragging = false,
  }) : super(key: key);

  final List<T> cardDeck;
  final Function() swipeLeft, swipeRight;
  final double minimumVelocity;
  final double rotationFactor;
  final double swipeThreshold;
  final double cardWidth;
  final Duration swipeAnimationDuration;
  final bool disableDragging;

  Alignment dragAlignment = Alignment.center;

  late final AnimationController swipeController;
  late Animation<Alignment> swipe;

  @override
  State<StatefulWidget> createState() => _SwipingGestureDetector();
}

class _SwipingGestureDetector extends State<SwipingGestureDetector>
    with TickerProviderStateMixin {
  bool animationActive = false;
  late final AnimationController springController;
  late Animation<Alignment> spring;

  @override
  void initState() {
    super.initState();
    springController = AnimationController(vsync: this);
    springController.addListener(() {
      setState(() {
        widget.dragAlignment = spring.value;
      });
    });

    widget.swipeController = AnimationController(
        vsync: this, duration: widget.swipeAnimationDuration);
    widget.swipeController.addListener(() {
      setState(() {
        widget.dragAlignment = widget.swipe.value;
      });
    });
  }

  @override
  void didUpdateWidget(covariant SwipingGestureDetector oldWidget) {
    super.didUpdateWidget(oldWidget);
    widget.swipeController = oldWidget.swipeController;
  }

  @override
  void dispose() {
    springController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final Size screenSize = MediaQuery.of(context).size;
    return GestureDetector(
      onPanUpdate: !widget.disableDragging ? _onPanUpdate : (_) {},
      onPanStart: !widget.disableDragging ? _onPanStart : (_) {},
      onPanEnd: !widget.disableDragging
          ? (details) => _onPanEnd(details, screenSize)
          : (_) {},
      child: Stack(
        alignment: Alignment.center,
        children: topTwoCards(),
      ),
    );
  }

  void _onPanUpdate(DragUpdateDetails details) {
    setState(() {
      widget.dragAlignment += Alignment(details.delta.dx, details.delta.dy);
    });
  }

  void _onPanStart(DragStartDetails details) async {
    if (animationActive) {
      springController.stop();
    }
  }

  void _onPanEnd(DragEndDetails details, Size screenSize) async {
    double vx = details.velocity.pixelsPerSecond.dx;
    if (vx >= widget.minimumVelocity ||
        widget.dragAlignment.x >= widget.swipeThreshold) {
      await widget.swipeRight();
    } else if (vx <= -widget.minimumVelocity ||
        widget.dragAlignment.x <= -widget.swipeThreshold) {
      await widget.swipeLeft();
    } else {
      animateBackToDeck(details.velocity.pixelsPerSecond, screenSize);
    }
    setState(() {
      widget.dragAlignment = Alignment.center;
    });
  }

  List<Widget> topTwoCards() {
    if (widget.cardDeck.isEmpty) {
      return [
        const SizedBox(
          height: 0,
          width: 0,
        )
      ];
    }
    List<Widget> cardDeck = [];
    int deckLength = widget.cardDeck.length;
    for (int i = max(deckLength - 2, 0); i < deckLength; ++i) {
      cardDeck.add(widget.cardDeck[i]);
    }
    Widget topCard = cardDeck.last;
    cardDeck.removeLast();
    cardDeck.add(
      Align(
          alignment: Alignment(getCardXPosition(), 0),
          child: Transform.rotate(
            angle: getCardAngle(),
            child: topCard,
          )),
    );
    return cardDeck;
  }

  double getCardAngle() {
    final double screenWidth = MediaQuery.of(context).size.width;
    return widget.rotationFactor * (widget.dragAlignment.x / screenWidth);
  }

  double getCardXPosition() {
    final double screenWidth = MediaQuery.of(context).size.width;
    return widget.dragAlignment.x / ((screenWidth - widget.cardWidth) / 2);
  }

  void animateBackToDeck(Offset pixelsPerSecond, Size size) async {
    spring = springController.drive(
      AlignmentTween(
        begin: widget.dragAlignment,
        end: Alignment.center,
      ),
    );

    // Calculate the velocity relative to the unit interval, [0,1],
    // used by the animation controller.
    final unitsPerSecondX = pixelsPerSecond.dx / size.width;
    final unitsPerSecondY = pixelsPerSecond.dy / size.height;
    final unitsPerSecond = Offset(unitsPerSecondX, unitsPerSecondY);
    final unitVelocity = unitsPerSecond.distance;

    const springProps = SpringDescription(
      mass: 30,
      stiffness: 1,
      damping: 1,
    );

    final simulation = SpringSimulation(springProps, 0, 1, -unitVelocity);
    animationActive = true;
    await springController.animateWith(simulation);
    animationActive = false;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:swiping_card_deck/swiping_card_deck.dart';
import 'dart:math' as math;

void main() {
  runApp(const MaterialApp(
    home: Scaffold(
        body: Center(
      child: ExamplePage(),
    )),
    title: 'SwipingCardDeck',
  ));
}

class ExamplePage extends StatelessWidget {
  const ExamplePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final SwipingCardDeck deck = SwipingCardDeck(
      cardDeck: getCardDeck(),
      onDeckEmpty: () => debugPrint("Card deck empty"),
      onLeftSwipe: (Card card) => debugPrint("Swiped left!"),
      onRightSwipe: (Card card) => debugPrint("Swiped right!"),
      cardWidth: 200,
      swipeThreshold: MediaQuery.of(context).size.width / 3,
      minimumVelocity: 1000,
      rotationFactor: 0.8 / 3.14,
      swipeAnimationDuration: const Duration(milliseconds: 500),
      disableDragging: false,
    );
    return Column(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        deck,
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          mainAxisSize: MainAxisSize.min,
          children: [
            IconButton(
              icon: const Icon(Icons.clear),
              iconSize: 30,
              color: Colors.red,
              onPressed: deck.animationActive ? null : () => deck.swipeLeft(),
            ),
            const SizedBox(width: 40),
            IconButton(
              icon: const Icon(Icons.check),
              iconSize: 30,
              color: Colors.green,
              onPressed: deck.animationActive ? null : () => deck.swipeRight(),
            ),
          ],
        ),
      ],
    );
  }

  List<Card> getCardDeck() {
    List<Card> cardDeck = [];
    for (int i = 0; i < 500; ++i) {
      cardDeck.add(
        Card(
            color: Color((math.Random().nextDouble() * 0xFFFFFF).toInt())
                .withOpacity(1.0),
            child: const SizedBox(height: 300, width: 200)),
      );
    }
    return cardDeck;
  }
}
